### ISSUE TITLE
The order of initContainers changed after the patch

### ISSUE THREAD

**Issue by OpenetCK** [kubernetes-sigs/kustomize#3912]
The initContainers  should maintain its order after the patching with kustomize.

**Files that can reproduce the issue**

kustomization.yaml
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - source.yaml
patches:
  - path: ./pod-patch.yaml
    target:
      kind: Pod
      name: ".*"

```

pod-patch.yaml
```
apiVersion: apps/v1
kind: Pod
metadata:
  name: doesNotMatter
spec:
  initContainers:
    - name: init-myservice
      env:
        - name: HTTP_ADDR
          value: https://[$(HOST_IP)]:8501
    - name: init-mydb
      env:
        - name: HTTP_ADDR
          value: https://[$(HOST_IP)]:8501
```

source.yaml
```
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
    - name: myapp-container
      image: busybox:1.28
      command: ['sh', '-c', 'echo The app is running! && sleep 3600']
  initContainers:
    - name: init-mydb
      image: busybox:1.28
      command: ['sh', '-c', "sleep 3600"]
    - name: init-myservice
      image: busybox:1.28
      command: ['sh', '-c', "sleep 7200"]


```

**Expected output**

The init container `init-mydb` should be first in the list.

**Actual output**

```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: myapp
  name: myapp-pod
spec:
  containers:
  - command:
    - sh
    - -c
    - echo The app is running! && sleep 3600
    image: busybox:1.28
    name: myapp-container
  initContainers:
  - command:
    - sh
    - -c
    - sleep 7200
    env:
    - name: HTTP_ADDR
      value: https://[$(HOST_IP)]:8501
    image: busybox:1.28
    name: init-myservice
  - command:
    - sh
    - -c
    - sleep 3600
    env:
    - name: HTTP_ADDR
      value: https://[$(HOST_IP)]:8501
    image: busybox:1.28
    name: init-mydb

```

**Kustomize version**

{Version:kustomize/v4.1.3 GitCommit:0f614e92f72f1b938a9171b964d90b197ca8fb68 BuildDate:2021-05-20T20:52:40Z GoOs:linux GoArch:amd64}

**Platform**

Linux 4.18.0-240.22.1.el8_3.x86_64 x86_64 GNU/Linux

**Additional context**

N/A

**Comment by monopole** [comment]
two questions:
* Did you try changing the order in the patch?
* Does the order of the initContainers in the result matter in some way we should be aware of?

**Comment by OpenetCK** [comment]
Yes, the order in the patch file did help to workaround, but it doesn't fix all especially when from the source file there are 3 init containers, but we only want to patch one of them, and remain the sequence.

Yes the order does matter for the initializing process, quoting from the kubernetes doc:

> If you specify multiple init containers for a Pod, kubelet runs each init container sequentially. Each init container must succeed before the next can run.

**Comment by serajin00** [comment]
Same problem happens on pod spec containers. The container order does change after strategic merge.
(tested with 4.1.3)

So it affects this kind of patches.
```
patches:
- target:
    kind: Deployment
  patch: |
    - op: add
      path: /spec/template/spec/containers/0/imagePullPolicy
      value: Always
```

**Comment by serajin00** [comment]
So quite a time passed in inactivity.

@monopole I guess this should be marked as kind/bug.

I as a [argocd](https://argoproj.github.io/argo-cd/) user, used argocd 1.6 which bundled with **Kustomize v3.6.1**.
I upgraded argocd 2.0 which bundled with **Kustomize v3.9.4**, then I surprised from the multibyte string error issue.
So I configured argocd to use the latest **Kustomize v4.1.3**, the multibyte issue fixed, but this issue happened. The order of container array changed after patch.
Currently, I stick to **Kustomize v3.6.1** because of this issue.

I as a Kustomize v3 user, this is a huge breaking change, I cannot upgrade to Kustomize v4.
Any array order should not be changed after patch.

**Comment by c0llab0rat0r** [comment]
I've run into this issue with the version of `kustomize` that's bundled with `kubectl`

> Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-17T02:23:27Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
> Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.4-3+adc4115d990346", GitCommit:"adc4115d990346b87714cc4f033d225711bf744d", GitTreeState:"clean", BuildDate:"2021-11-17T22:03:17Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}

If I had used `kubectl -k` it's not likely I would have caught the initContainers ordering problem. Luckily I was saving the output to a file first and reviewing it.

My situation is described in a prior comment. A pod in a template has two init containers, and only one of the init containers is referenced in the patch. `kustomize` renders the init containers out of order, breaking the pod.

/remove-lifecycle stale

**Comment by epcim** [comment]
I face the same issue.

Deploing an gitea Helm chart. See they do expect, the first initContainer will format `app.ini` config first: https://gitea.com/gitea/helm-chart/src/branch/master/templates/gitea/statefulset.yaml#L75

tl;dr - `init-app-ini` container must run before `configure-gitea`

Corresponding part in my Customize:
```
resources:
- ./secret/sealed-gitea-app-ini-database.yaml
- ./secret/sealed-gitea-admin.yaml

# patch gitea, this touch the intiContainer
patchesStrategicMerge:
- overlays/patchGitea.yaml. <<-- touching env variables
```


I have tried reorder option, generation to individual files, single output - as long as I patch the given initContainer Kustomize will return wrong order.

**Comment by severinraez** [comment]
FWIW, the workaround I'm using for now (v4.4.0) is to reference the initContainers in their correct order in the patch, but not change them

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: rails
spec:
  template:
    spec:
      initContainers:
        - name: wait-for-database # required so order does not change
        - name: migrate-database
          command:
            - "/usr/bin/bash"
#            - "-c"
            - "bundle exec rake bleib:wait_for_migrations"
[...]
```

**Comment by severinraez** [comment]
/remove-lifecycle rotten

Kubernetes Triage Robot ***@***.***> schrieb am Do., 26. Mai
2022, 16:54:

**Comment by davinkevin** [comment]
We had the same error with `kustomize` v4.5.7 and v5.0.1, and choose to use the same solution as @severinraez from [here](https://github.com/kubernetes-sigs/kustomize/issues/3912#issuecomment-1022218731)

**Comment by shapirus** [comment]
This issue should be reopened. The bug still exists. The workaround is not suitable when the names of other containers than that being patched are not known (e.g. having the patches defined in a Component reused for multiple sets of different manifests).
